src/compiler/crystal/semantic/type_guess_visitor.cr

```diff
@@ -601,6 +601,32 @@ module Crystal
       end
     end
 
+    def guess_type(node : ProcLiteral)
+      output = node.def.return_type
+      return nil unless output
```

Comment:
I really just don't understand this code. Why are we seemingly refusing to guess the type in this case? Wouldn't we need to guess the type if there is *no* annotation, rather than when there is one? (as I understand, `node.def.return_type` is the annotation)

Reply:
`Crystal::TypeGuessVisitor` is for instance / class variables (i.e. the `getter foo` in the OP); if there isn't an output type restriction, we cannot type the body here yet and we cannot instantiate `Proc` with `R = _` either. (Maybe we can nonetheless guess the type from simple bodies but that's for later.)